### PR TITLE
llvm-r1.eclass: new fancy eclass with `LLVM_SLOT` `USE_EXPAND`

### PR DESCRIPTION
**Pull request by mgorny**
@gentoo/llvm

### CODE REVIEW

**Review by ArsenArsen — APPROVED** [review]
this seems okay at a glance, thanks

File: eclass/llvm-r1.eclass [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+		# prepend new path in front of the first LLVM version found
+		for x in "${split_path[@]}"; do
+			if [[ ${x} == */usr/lib/llvm/*/bin ]]; then
+				if [[ ${x} != ${llvm_path} ]]; then
+					new_path+=( "${llvm_path}" )
+				elif [[ ${added} && ${x} == ${llvm_path} ]]; then
+					# deduplicate
+					continue
+				fi
+				added=1
```
Comment (miller-alex) [inline eclass/llvm-r1.eclass:299]: This adds `${llvm_path}` in front of every other LLVM version, not just the first.
The following code should work:
```
                        if [[ ! ${added} ]]; then
                                new_path+=( "${llvm_path}" )
                                added=1
                        fi
                        [[ ${x} == ${llvm_path} ]] && continue # deduplicate
```

Comment (mgorny) [reply eclass/llvm-r1.eclass:299]: Thanks. Indeed it does. I'm going to fix it, and add tests for all functions.

Curious enough, the code is copied from `llvm.eclass` and apparently the test suite there wasn't good enough.

File: dev-util/intel_clc/intel_clc-24.0.0.ebuild [hunk 2]
```diff
@@ -41,53 +47,17 @@ BDEPEND="
...
 
 pkg_setup() {
-	llvm_pkg_setup
+	llvm-r1_pkg_setup
 	python-any-r1_pkg_setup
 }
 
 src_configure() {
-	PKG_CONFIG_PATH="$(get_llvm_prefix "${LLVM_MAX_SLOT}")/$(get_libdir)/pkgconfig"
+	PKG_CONFIG_PATH="${ESYSROOT}$(get_llvm_prefix)/$(get_libdir)/pkgconfig"
```
Comment (mgorny) [inline dev-util/intel_clc/intel_clc-24.0.0.ebuild:60]: I'm wondering if I should revert to including `EPREFIX`/`BROOT` (via option arg) in this, as I suspect people will fail to prepend the right prefix.

**Comment by mgorny** [comment]
I've split path prepending into a separate function, and added a test suite.

File: dev-util/intel_clc/intel_clc-24.0.0.ebuild (hunk 2 above)
Comment (mgorny) [reply dev-util/intel_clc/intel_clc-24.0.0.ebuild]: Done that now.

**Comment by mgorny** [comment]
Now with a split `llvm-utils.eclass` that's shared between the old and new eclasses, and will be used directly by some LLVM ebuilds.

**Review by ArsenArsen — APPROVED** [review]
eclass changes look ok at a glance

File: eclass/llvm-utils.eclass [hunk 3]
```diff
@@ -132,7 +132,7 @@ llvm_prepend_path() {
 	# prepend new path in front of the first LLVM version found
 	for x in "${split_path[@]}"; do
 		if [[ ${x} == */usr/lib/llvm/*/bin ]]; then
-			if [[ ${x} != ${llvm_path} ]]; then
+			if [[ ${x} != ${llvm_path} && ! ${added} ]]; then
```
Comment (miller-alex) [inline eclass/llvm-utils.eclass]: That should work, although I find my other suggestion easier to understand.

I really wanted to comment on the commit message:
> Thanks to @miller-alex who spotted it in:
#35196 (comment)

I never use that github account name anywhere, I'd prefer `Alexander Miller <alex.miller@gmx.de>` (or just the name) like all my other contributions.

File: profiles/arch/base/use.mask [hunk 4]
```diff
@@ -1,6 +1,11 @@
 # Copyright 1999-2024 Gentoo Authors
 # Distributed under the terms of the GNU General Public License v2
 
+# Michał Górny <mgorny@gentoo.org> (2021-02-05)
```
Comment (miller-alex) [inline profiles/arch/base/use.mask:4]: It's 2024 already.
Same issue in `profiles/arch/loong/use.mask`, line 4.

File: eclass/llvm-r1.eclass [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+# An eclass to reliably depend on a set of LLVM-related packages
+# in a matching slot.  To use the eclass:
+#
+# 1. Set LLVM_COMPAT to the list of supported LLVM slots.
+# 2. Use llvm_gen_dep and/or LLVM_USEDEP to add appropriate
+#    dependencies.
+# 3. Use llvm-r1_pkg_setup, get_llvm_prefix or LLVM_SLOT.
+#
+# The eclass sets IUSE and REQUIRED_USE.  The flag corresponding
+# to the newest supported stable LLVM slot is enabled by default.
```
Comment (miller-alex) [inline eclass/llvm-r1.eclass:22]: Nit-picking. That's only true if (1) `LLVM_COMPAT` is properly sorted (which it should be according to the description, but who knows) and (2) it contains a stable slot. Why not admit that it simply picks the last (stable, if available) slot.

File: eclass/llvm-r1.eclass [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+# be automatically enabled via USE defaults.
+_LLVM_NEWEST_STABLE=17
+
+# == control variables ==
+
+# @ECLASS_VARIABLE: LLVM_COMPAT
+# @PRE_INHERIT
+# @REQUIRED
+# @DESCRIPTION:
+# A list of LLVM slots supported by the package, oldest to newest.
```
Comment (miller-alex) [inline eclass/llvm-r1.eclass:71]: Is the order relevant except for picking the latest by default? (See other comment above)

File: eclass/llvm-r1.eclass [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+}
+
+# == ebuild helpers ==
+
+# @FUNCTION: get_llvm_prefix
+# @USAGE: [-b|-d]
+# @DESCRIPTION:
+# Output the path to the selected LLVM slot.  Note that this is path
+# relative to the prefix, i.e. you need to prepend ${ESYSROOT}
+# or ${BROOT} appropriately.
```
Comment (miller-alex) [inline eclass/llvm-r1.eclass:165]: The second sentence looks like an obsolete remnant from before the addition of the `-b` and `-d` options.
It should be removed.

**Review by thesamesam — APPROVED** [review]
Please tag the following bugs where appropriate:
* https://bugs.gentoo.org/923228 (motivation/what pushed us to finally do this)
* https://bugs.gentoo.org/880671 / https://bugs.gentoo.org/821955 case, as I think this is going to end up solving that too, actually.
* https://bugs.gentoo.org/919150 (fixes this/makes it obsolete)

I think I'm happy overall, but I'd like to maybe see some more ebuilds ported first to make sure we're not missing anything.

**Comment by thesamesam** [comment]
@juippis It just occurred to me that we'd benefit from knowing if FF needs any API changes too. Could you have a look to see if the eclass as proposed meets the mozilla stack's needs? thanks!

File: eclass/llvm-r1.eclass (hunk 5 above)
Comment (mgorny) [reply eclass/llvm-r1.eclass:22]: I don't want people making assumptions that aren't guaranteed. That said, this comment is a bit outdated, so I've updated it to include the possibility of enabling testing version by default.

File: eclass/llvm-r1.eclass (hunk 6 above)
Comment (mgorny) [reply eclass/llvm-r1.eclass:71]: It might be in the future.

File: eclass/llvm-r1.eclass (hunk 7 above)
Comment (mgorny) [reply eclass/llvm-r1.eclass:165]: Good catch, thanks!

File: eclass/llvm-utils.eclass (hunk 3 above)
Comment (mgorny) [reply eclass/llvm-utils.eclass]: I think your suggestion actually could result in duplicates if first entry in the path matches the selected path. But yeah, the logic is kinda ugly, I need to take a second look at it.

Comment (mgorny) [reply eclass/llvm-utils.eclass]: Actually, you were right. I've added some more test cases, and actually arrived at the same code. Thanks again!

**Comment by mgorny** [comment]
Updated per review comments.

**Comment by ArsenArsen** [comment]
```
4e5c02cf43bf /var/db/repos/gentoo # cat /var/db/pkg/dev-util/intel_clc-24.0.0/RDEPEND 
dev-libs/libclc dev-util/spirv-tools >=sys-libs/zlib-1.2.8:0/1= x11-libs/libdrm dev-util/spirv-llvm-translator:17 sys-devel/clang:17 sys-devel/llvm:17 >=sys-libs/glibc-2.38-r9
```
can also confirm that the version push you just did removed the incorrect dependency in a test container \o/

thanks again

**Comment by thesamesam** [comment]
'See-Also' in Gentoo is `Bug:` ;)

Anyway, thanks a lot for working on this. Let's wait if we can a little bit for juippis to try Firefox, but if you want to go ahead and we just take the risk of adapting stuff later, that works for me too.

**Comment by juippis** [comment]
```
>>> Emerging (127 of 127) www-client/firefox-122.0.1::gentoo
>>> Installing (127 of 127) www-client/firefox-122.0.1::gentoo
>>> Completed (127 of 127) www-client/firefox-122.0.1::gentoo
```
Sweet, at least basic emerge went through.
`www-client/firefox-122.0.1:rapid::gentoo  USE="X clang dbus gmp-autoupdate jumbo-build system-av1 system-harfbuzz system-icu system-jpeg system-libevent system-libvpx system-webp telemetry -debug -eme-free -geckodriver -hardened -hwaccel -jack -libproxy -lto -openh264 -pgo -pulseaudio (-selinux) -sndio -system-png (-system-python-libs) (-valgrind) -wayland -wifi" L10N="fi sv -ach -af -an -ar -ast -az -be -bg -bn -br -bs -ca -ca-valencia -cak -cs -cy -da -de -dsb -el -en-CA -en-GB -eo -es-AR -es-CL -es-ES -es-MX -et -eu -fa -ff -fr -fur -fy -ga -gd -gl -gn -gu -he -hi -hr -hsb -hu -hy -ia -id -is -it -ja -ka -kab -kk -km -kn -ko -lij -lt -lv -mk -mr -ms -my -nb -ne -nl -nn -oc -pa -pl -pt-BR -pt-PT -rm -ro -ru -sc -sco -si -sk -sl -son -sq -sr -szl -ta -te -th -tl -tr -trs -uk -ur -uz -vi -xh -zh-CN -zh-TW" LLVM_SLOT="17 -16"`

This probably allows to clean tons of local `llvm` hacks, but will tend to those naturally on version bumps.

```diff
---
 www-client/firefox/firefox-122.0.1.ebuild | 35 +++++++++--------------
 1 file changed, 13 insertions(+), 22 deletions(-)

diff --git a/www-client/firefox/firefox-122.0.1.ebuild b/www-client/firefox/firefox-122.0.1.ebuild
index c5c84ee9f..f443d84b3 100644
--- a/www-client/firefox/firefox-122.0.1.ebuild
+++ b/www-client/firefox/firefox-122.0.1.ebuild
@@ -5,7 +5,7 @@ EAPI=8
 
 FIREFOX_PATCHSET="firefox-122-patches-02.tar.xz"
 
-LLVM_MAX_SLOT=17
+LLVM_COMPAT=( 16 17 )
 
 PYTHON_COMPAT=( python3_{10..11} )
 PYTHON_REQ_USE="ncurses,sqlite,ssl"
@@ -37,7 +37,7 @@ MOZ_P="${MOZ_PN}-${MOZ_PV}"
 MOZ_PV_DISTFILES="${MOZ_PV}${MOZ_PV_SUFFIX}"
 MOZ_P_DISTFILES="${MOZ_PN}-${MOZ_PV_DISTFILES}"
 
-inherit autotools check-reqs desktop flag-o-matic gnome2-utils linux-info llvm multiprocessing \
+inherit autotools check-reqs desktop flag-o-matic gnome2-utils linux-info llvm-r1 multiprocessing \
        optfeature pax-utils python-any-r1 readme.gentoo-r1 toolchain-funcs virtualx xdg
 
 MOZ_SRC_BASE_URI="https://archive.mozilla.org/pub/${MOZ_PN}/releases/${MOZ_PV}"
@@ -81,25 +81,16 @@ FF_ONLY_DEPEND="!www-client/firefox:0
        !www-client/firefox:esr
        selinux? ( sec-policy/selinux-mozilla )"
 BDEPEND="${PYTHON_DEPS}
-       || (
-               (
-                       sys-devel/clang:17
-                       sys-devel/llvm:17
-                       clang? (
-                               sys-devel/lld:17
-                               virtual/rust:0/llvm-17
-                               pgo? ( =sys-libs/compiler-rt-sanitizers-17*[profile] )
-                       )
-               )
-               (
-                       sys-devel/clang:16
-                       sys-devel/llvm:16
-                       clang? (
-                               sys-devel/lld:16
-                               virtual/rust:0/llvm-16
-                               pgo? ( =sys-libs/compiler-rt-sanitizers-16*[profile] )
-                       )
-               )
+       $(llvm_gen_dep '
+               sys-devel/clang:${LLVM_SLOT}
+               sys-devel/llvm:${LLVM_SLOT}
+       ')
+       clang? (
+               $(llvm_gen_dep '
+                       sys-devel/lld:${LLVM_SLOT}
+                       virtual/rust:0/llvm-${LLVM_SLOT}
+               ')
+               $(llvm_gen_dep 'pgo? ( sys-libs/compiler-rt-sanitizers:${LLVM_SLOT}[profile] )')
        )
        app-alternatives/awk
        app-arch/unzip
@@ -521,7 +512,7 @@ pkg_setup() {
 
                check-reqs_pkg_setup
 
-               llvm_pkg_setup
+               llvm-r1_pkg_setup
 
                if use clang && use lto && tc-ld-is-lld ; then
                        local version_lld=$(ld.lld --version 2>/dev/null | awk '{ print $2 }')
-- 
2.43.0

```
(proof-of-concept)

**Comment by mgorny** [comment]
Ideally, put everything inside a single `llvm_gen_dep` block, that's costly.

**Comment by juippis** [comment]
Yep, that's cleaner and should work. Will test but don't wait for me to merge, it's proven to work at least adequately already.

File: eclass/llvm-r1.eclass [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+# DEPEND="
+#   $(llvm_gen_dep '
+#     sys-devel/clang:${LLVM_SLOT}
+#     sys-devel/llvm:${LLVM_SLOT}
+#   ')
+# "
+# @CODE
+llvm_gen_dep() {
+	debug-print-function ${FUNCNAME} "${@}"
+
```
Comment (ionenwks) [inline eclass/llvm-r1.eclass:148]: It's not shown on github but there's an unnecessary tab on the blank line (finding the real problems).

**Review by ionenwks — COMMENTED** [review]
Tried migrating qt-creator and qttools locally (trivial) and no issues. Unsure when planning to merge, but if done before Qt6.6.2 releases (next week or so) I'll update these at same time to spare people extra rebuilds.

**Comment by ionenwks** [comment]
Wonder if it'd make sense to have a way for REQUIRED_USE to be optional so that a user could turn them all off to avoid rebuilds when `USE="-clang/llvm"`.

With python it's not unusual to have `REQUIRED_USE="python? ( ${PYTHON_REQUIRED_USE} )"`

Could live without extra complexity though, just came to mind given both qttools and qt-creator only use it with USE=clang and the LLVM_SLOTs are otherwise no-ops. Change in IUSE defaults (new slots) will cause rebuilds even if unused but ah well.

File: eclass/llvm-r1.eclass (hunk 8 above)
Comment (mgorny) [reply eclass/llvm-r1.eclass:148]: Good catch, thanks! Fixed locally.

**Comment by mgorny** [comment]
Well, I suppose we could do that but I don't think that would change much. The main problem is in defaults, and we can't make them conditional. And people who go all the way to disable the flags to prevent rebuilds, can as well set them to a fixed target.

**Comment by ionenwks** [comment]
Yeah, you're right, may as well live with it. Defaults won't change that often anyway.

**Comment by mgorny** [comment]
> Yeah, you're right, may as well live with it. Defaults won't change that often anyway.

Damn you, now that you've made me realize it, I won't be able to leave it alone, so `LLVM_OPTIONAL` is coming. As a side effect, this will make it possible to utilize `IUSE_RUNTIME` or something else, in the future, to "mask" these flags when they have no effect.

**Comment by mgorny** [comment]
Now with `LLVM_OPTIONAL`.

**Review by ArsenArsen — APPROVED** [review]

**Review by ionenwks — APPROVED** [review]
Migrated qttools and qt-creator to LLVM_OPTIONAL locally and all good. Thanks for that, may not change much but it still feels righter.

I'd say may as well just merge whenever so it can get wider testing.

**Review by miller-alex — APPROVED** [review]

**Review by thesamesam — APPROVED** [review]

**Pull request closed without merging**